enhancements/microshift/microshift-apiserver-certs.md

```diff
@@ -0,0 +1,312 @@
+
+When creating the API server instance there is a [default certificate](https://github.com/openshift/microshift/blob/main/pkg/controllers/kube-apiserver.go#L136)
+which is the serving service network. The rest of the certificates are
+configured [here](https://github.com/openshift/microshift/blob/main/pkg/controllers/kube-apiserver.go#L184-L202)
+along with the TLS versions and cipher suites. These go into an upstream
+library called `dynamiccertificates`. If certificates have names on them (which
+they do), the package [creates](https://github.com/kubernetes/apiserver/blob/master/pkg/server/dynamiccertificates/tlsconfig.go#L209)
+a 1:1 mapping of names to certificates. There is one entry per name in each
+certificate, including Common Name (CN) and Subject Alternative Names (SAN).
+Note there can only be one certificate per name, as otherwise there would be
+ambiguity.
+
+#### How API server selects certificates for incoming connections
+Everytime there is an incoming connection to the API server dynamiccertificates
+package kicks in to select the certificate it should serve. If the client sends
+SNI information (which means a name has been used, instead of an IP), then TLS
+configuration follows normal flow. If SNI was not used, then it will start a
+lookup in the table we described above using the destination IP for the
+connection. This destination IP is one where API server is listening. If there
+is no match (meaning that no certificate was configured with CN/SAN to include
+the API server IP), then it resolves as if it were SNI, which will end up
+returning the default certificate.
+
+#### Connecting to API server
+Here we can distinguish two use cases: external access and internal access.
+
+External access will use the kubeconfig, which contains a `server` stanza in it
+with the name (DNS or IP) and port of the cluster. As seen in previous section,
+if its a name it will find a matching certificate, and if its an IP it will try
+looking it up in the mapping table, returning the default certificate if not
+present. Note the name or IP must be in the external certificate, as there is
+hostname validation on the client.
+
+Internal access is typically using [client-go](https://github.com/kubernetes/client-go).
+If using [InClusterConfig](https://github.com/kubernetes/client-go/blob/master/rest/config.go#L511)
+then it will not use the name for connecting. As we can see it uses the IP for
+the `kubernetes` service, injected through the environment variables. This IP
+is the service network IP, which later on gets translated to an endpoint:
+```bash
+$ KUBECONFIG=/var/lib/microshift/resources/kubeadmin/kubeconfig oc get svc kubernetes -o yaml
+apiVersion: v1
+kind: Service
+metadata:
+  creationTimestamp: "2022-12-20T15:57:23Z"
+  labels:
+    component: apiserver
+    provider: kubernetes
+  name: kubernetes
+  namespace: default
+  resourceVersion: "199"
+  uid: 28e299a6-1578-4fd0-babd-660a1a86cdfb
+spec:
+  clusterIP: 10.43.0.1
```

Comment:
Kind of a digression, and I suspect that this wouldn't work because the kubernetes service is "special" (internally hard-coded, etc.), but I wonder if things would be simplified and/or perform better if it were configured as a headless service (cluster IP is not allocated, no proxying).

Reply:
The catch is that the apiserver does not have a selector as there is no deployment. It is using a special label in the endpoints so that the IP is injected:
```bash
# KUBECONFIG=/var/lib/microshift/resources/kubeadmin/kubeconfig oc get svc,endpointslices kubernetes -o yaml
apiVersion: v1
items:
- apiVersion: v1
  kind: Service
  metadata:
    creationTimestamp: "2023-01-23T14:06:35Z"
    labels:
      component: apiserver
      provider: kubernetes
    name: kubernetes
    namespace: default
    resourceVersion: "198"
    uid: 2c1a3f64-345a-4c46-9025-812fe1c5d442
  spec:
    clusterIP: 10.43.0.1
    clusterIPs:
    - 10.43.0.1
    internalTrafficPolicy: Cluster
    ipFamilies:
    - IPv4
    ipFamilyPolicy: SingleStack
    ports:
    - name: https
      port: 443
      protocol: TCP
      targetPort: 6443
    sessionAffinity: None
    type: ClusterIP
  status:
    loadBalancer: {}
- addressType: IPv4
  apiVersion: discovery.k8s.io/v1
  endpoints:
  - addresses:
    - 192.168.122.117
    conditions:
      ready: true
  kind: EndpointSlice
  metadata:
    creationTimestamp: "2023-01-23T14:06:35Z"
    generation: 1
    labels:
      kubernetes.io/service-name: kubernetes
    name: kubernetes
    namespace: default
    resourceVersion: "201"
    uid: 7daf3fee-99f5-4bbd-a92c-a36e9944dcd7
  ports:
  - name: https
    port: 6443
    protocol: TCP
kind: List
metadata:
  resourceVersion: ""
```